lighthouse-core/lib/page-functions.js

```diff
@@ -186,7 +186,11 @@ function getNodePath(node) {
   while (node && node.parentNode) {
     const index = getNodeIndex(node);
     path.push([index, node.nodeName]);
-    node = node.parentNode;
+    if (node.parentNode.nodeType === Node.DOCUMENT_FRAGMENT_NODE) {
```

Comment:
would just this work:

```
node = (node.parentNode || {}).host || node.parentNode
```

(excuse the terseness in code)

Reply:
`node.parentNode` is already checked by the loop conditional so should be able to do just `node = node.parentNode.host || node.parentNode;`, but is a "`host` exists" check sufficient? (maybe, but people do weird things to their DOMs)

Could also go more elaborate and do
```js
if (node.parentNode.host && node.parentNode.host.shadowRoot) {
  // Step out of shadow DOM.
  node = node.parentNode.host;
} else {
  node = node.parentNode;
}
```
(or get really fancy and do `node.parentNode.host.shadowRoot === node.parentNode` for the second part of the conditional, but that might be overkill :)

Reply:
i'll stop ya'll right now since this PR doesn't work. :)

Reply:
Well that's good to know, a little more information would be useful. 

I'll add the change, and @brendankenny the current node at this point would be the shadow root so I'm not sure why you would check that the host's root is the node you're on. How can you have a host who isn't your host?

Reply:
@paulirish i tried this myself on chromestatus' page, on one of the paragraphs just like in your issue. I got this path:

```
2,HTML,1,BODY,0,APP-DRAWER-LAYOUT,1,APP-HEADER-LAYOUT,1,DIV,1,CHROMEDASH-FEATURELIST,154,DIV,1,CHROMEDASH-FEATURE,2,SECTION,0,SUMMARY,1,P
```

what part of it doesn't work? is it that you expected something else in the path? please elaborate so this can be fixed

Reply:
> I'll add the tests, and @brendankenny the current node at this point would be the shadow root so I'm not sure why you would check that the host's root is the node you're on. AFAIK a document fragment can't be a host unless it has a shadowRoot anyway.

Right, I was saying that just checking `host` exists isn't enough because it isn't a guarded property for most `Node`s (just `ShadowRoot`s), e.g. you could set `node.host = 'some string'` on non-ShadowRoots, and if you were just checking for `host`, then the path traversal would stop at that node.

Your solution wouldn't have that problem, though as you said, it does feel a little weird to be checking for a `DocumentFragment` and relying on the fact that `DocumentFragment`s can't have a `parentNode` (but they can have a `host` if they're also a `ShadowRoot`).

My suggestion is a little more explicit in looking for something with a host that has a ShadowRoot, but if we feel confident new uses of DocumentFragments won't break our assumption in the future, a comment explaining the situation might also be sufficient for future people editing this code.

Reply:
> @paulirish i tried this myself on chromestatus' page, on one of the paragraphs just like in your issue.

I'm curious too. Is it that the path doesn't indicate that you need to traverse into the shadow DOM (and not continue into the light DOM)? I noticed when you copy the full XPath in the Elements panel you get a `//` when going into ShadowRoot or DocumentFragment (like in a `template`), but I'm not sure how that translates into our comma-based strings. Double comma?

Reply:
i did add a small change so the path contains the shadow root as an entry, too. maybe that was it, we shall see...

and yes thats a great point, someone could set `host` to anything on a document fragment. its very unlikely but not impossible. the ideal solution is simply `instanceof ShadowRoot` but i didn't want to do that in case there's a situation `window` (in context of the function call) is different to the window of the node...

Reply:
ah, looks like it does need the explicit shadow root, yes. @paulirish had an example in https://github.com/GoogleChrome/lighthouse/issues/10914#issue-631320251 (though with `#document-fragment`...I assume it will be `#shadow-root` for this)

Reply:
> I assume it will be `#shadow-root` for this

no, ignore that, apparently it should be `#document-fragment` :)

Reply:
yep, the current function gives me this:

```
2,HTML,1,BODY,0,APP-DRAWER-LAYOUT,1,APP-HEADER-LAYOUT,1,DIV,1,CHROMEDASH-FEATURELIST,0,#document-fragment,164,DIV,1,CHROMEDASH-FEATURE,0,#document-fragment,2,SECTION,0,SUMMARY,1,P
```